### PR TITLE
ISSUE-209 LongLivedToken repository (API, memory, contract)

### PR DESCRIPTION
**Pull request by vttranlina**
ref https://github.com/linagora/tmail-backend/issues/209

### CODE REVIEW

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/longlivedtoken/LongLivedTokenStore.scala [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+case class LongLivedTokenNotFoundException() extends RuntimeException
+
+trait LongLivedTokenStore {
+  def store(username: Username, longLivedToken: LongLivedToken): Publisher[LongLivedTokenId]
+
+  def validate(username: Username, secret: LongLivedTokenSecret): Publisher[LongLivedTokenFootPrint]
+
+  def listTokens(user: Username): Publisher[LongLivedTokenFootPrint]
+
+  def revoke(username: Username, id: LongLivedTokenFootPrint): Publisher[Unit]
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/longlivedtoken/LongLivedTokenStore.scala:48]: Can't we just take (username, id) here?

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/longlivedtoken/LongLivedTokenStore.scala [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+}
+
+class InMemoryLongLivedTokenStore(tokenStore: Map[Username, Seq[ImmutableTriple[DeviceId, LongLivedTokenId, LongLivedTokenSecret]]] = scala.collection.concurrent.TrieMap())
+  extends LongLivedTokenStore {
+
+  override def store(username: Username, longLivedToken: LongLivedToken): Publisher[LongLivedTokenId] = {
+    Preconditions.checkNotNull(username)
+    Preconditions.checkNotNull(longLivedToken)
+    SMono.just(tokenStore.contains(username))
+      .filter(userExists => userExists)
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/longlivedtoken/LongLivedTokenStore.scala:58]: I feel it is a `if` in disguise...

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/longlivedtoken/LongLivedTokenStore.scala [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+  def store(username: Username, longLivedToken: LongLivedToken): Publisher[LongLivedTokenId]
+
+  def validate(username: Username, secret: LongLivedTokenSecret): Publisher[LongLivedTokenFootPrint]
+
+  def listTokens(user: Username): Publisher[LongLivedTokenFootPrint]
+
+  def revoke(username: Username, id: LongLivedTokenFootPrint): Publisher[Unit]
+}
+
+class InMemoryLongLivedTokenStore(tokenStore: Map[Username, Seq[ImmutableTriple[DeviceId, LongLivedTokenId, LongLivedTokenSecret]]] = scala.collection.concurrent.TrieMap())
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/longlivedtoken/LongLivedTokenStore.scala:51]: You should initialise the map within the class as a `val` and not leak the underlying map in the constructor to the outer world.

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/longlivedtoken/LongLivedTokenStore.scala [hunk 4]
```diff
@@ -0,0 +1,111 @@
+package com.linagora.tmail.james.jmap.longlivedtoken
+
+import com.google.common.base.Preconditions
+import org.apache.commons.lang3.tuple.ImmutableTriple
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/longlivedtoken/LongLivedTokenStore.scala:4]: I bet scala have native triple. Also we could have case classes to cover the usage of this... 

(Maybe using case classes is better)

File: tmail-backend/jmap/extensions/src/test/scala/com/linagora/tmail/james/jmap/longlivedtoken/LongLivedTokenStoreContract.scala [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+  @Test
+  def storeShouldThrowWhenLongLivedTokenIsNull(): Unit = {
+    assertThatThrownBy(() => SMono.fromPublisher(testee.store(USERNAME, null)).block())
+      .isInstanceOf(classOf[NullPointerException])
+  }
+
+  @Test
+  def storeShouldReturnLongLivedTokenId(): Unit = {
+    assertThat(SMono.fromPublisher(testee.store(USERNAME, LONG_LIVED_TOKEN)).block())
+      .isInstanceOf(classOf[LongLivedTokenId])
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions/src/test/scala/com/linagora/tmail/james/jmap/longlivedtoken/LongLivedTokenStoreContract.scala:45]: Calling that on two different token should return 2 different tokenIds?

File: tmail-backend/jmap/extensions/src/test/scala/com/linagora/tmail/james/jmap/longlivedtoken/LongLivedTokenStoreContract.scala [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+
+  @Test
+  def validateShouldReturnLongLivedTokenFootPrintWhenSecretValid(): Unit = {
+    val longLivedTokenId: LongLivedTokenId = SMono.fromPublisher(testee.store(USERNAME, LONG_LIVED_TOKEN)).block()
+    assertThat(SMono.fromPublisher(testee.validate(USERNAME, LONG_LIVED_TOKEN_SECRET)).block())
+      .isEqualTo(LongLivedTokenFootPrint(longLivedTokenId, DEVICE_ID))
+  }
+
+  @Test
+  def validateShouldThrowWhenLongLivedTokenSecretInvalid(): Unit = {
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions/src/test/scala/com/linagora/tmail/james/jmap/longlivedtoken/LongLivedTokenStoreContract.scala:90]:
```suggestion
  def validateShouldThrowWhenLongLivedTokenBelong to anotherUser(): Unit = {
```

File: tmail-backend/jmap/extensions/src/test/scala/com/linagora/tmail/james/jmap/longlivedtoken/LongLivedTokenStoreContract.scala [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+  }
+
+  @Test
+  def revokeShouldThrowWhenLongLivedTokenFootPrintIsNull(): Unit = {
+    assertThatThrownBy(() => SMono.fromPublisher(testee.revoke(USERNAME, null)).block())
+      .isInstanceOf(classOf[NullPointerException])
+  }
+
+  @Test
+  def revokeShouldThrowWhenLongLivedTokenFootPrintNotFound(): Unit = {
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions/src/test/scala/com/linagora/tmail/james/jmap/longlivedtoken/LongLivedTokenStoreContract.scala:155]: That is not indempotent. IMO we should not throw here.

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/longlivedtoken/LongLivedTokenStore.scala (hunk 4 above)
Comment (vttranlina) [reply tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/longlivedtoken/LongLivedTokenStore.scala:4]: yes, I see  Tuple3, But I think with `ImmutableTriple` it is like as "readOnly"

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/longlivedtoken/LongLivedTokenStore.scala (hunk 1 above)
Comment (vttranlina) [reply tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/longlivedtoken/LongLivedTokenStore.scala:48]: Do you mean is we don't need "DeviceId" for this method?

Comment (chibenwa) [reply tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/longlivedtoken/LongLivedTokenStore.scala:48]: I mean the LongLivedTokenId

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/longlivedtoken/LongLivedTokenStore.scala (hunk 4 above)
Comment (chibenwa) [reply tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/longlivedtoken/LongLivedTokenStore.scala:4]: In scala you can write `(a, b, c)`

But I prefer case classes

**Review by chibenwa — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]

**Pull request merged**